server/src/main/java/org/elasticsearch/common/xcontent/XContentHelper.java

```diff
@@ -748,4 +748,24 @@ public static XContentParser mapToXContentParser(XContentParserConfiguration con
             throw new ElasticsearchGenerationException("Failed to generate [" + source + "]", e);
         }
     }
+
+    public static Tuple<XContentParser, XContentParser> cloneParser(XContentParser parser) throws IOException {
+        XContentBuilder builder = XContentBuilder.builder(parser.contentType().xContent());
+        builder.copyCurrentStructure(parser);
+
+        var bytes = BytesReference.bytes(builder);
```

Comment:
Another option is to return the bytes and the parser, instead of two parsers. The bytes can be directly added to _ignored_source through `XContentHelper.TypeUtils.encode`. This avoids the duplicate call to `XContentBuilder.builder(parser.contentType().xContent())` in `TypeUtils.START.encode` for qualifying objects - I'd think these are heavyweight calls. Or, you can provide this specialized implementation for `START_OBJECT - START_ARRAY` tokens only.

Reply:
btw I made some progress with refactoring and tests on this part in https://github.com/elastic/elasticsearch/pull/108417